File: tests/analytics/tdee.test.ts
import { describe, test, expect } from 'vitest';
import { computeAdaptiveTDEE, detectPlateau, projectWeight } from '$lib/analytics/tdee';

function makeWeightSeries(days: number, startKg: number, dailyChange: number) {
	const today = new Date();
	today.setUTCHours(0, 0, 0, 0);
	return Array.from({ length: days }, (_, i) => {
		const d = new Date(today);
		d.setUTCDate(d.getUTCDate() - (days - 1 - i));
		return {
			date: d.toISOString().slice(0, 10),
			weightKg: startKg + i * dailyChange
		};
	});
}

function makeCalorieSeries(days: number, kcal: number) {
	const today = new Date();
	today.setUTCHours(0, 0, 0, 0);
	return Array.from({ length: days }, (_, i) => {
		const d = new Date(today);
		d.setUTCDate(d.getUTCDate() - (days - 1 - i));
		return {
			date: d.toISOString().slice(0, 10),
			calories: kcal
		};
	});
}

describe('computeAdaptiveTDEE', () => {
	test('estimates TDEE for a weight loss scenario', () => {
		// 0.5 kg/week loss = ~71g/day, 7700 kcal/kg → deficit ~550 kcal/day
		// intake 2000 → TDEE ≈ 2550
		const weeklyLossKg = 0.5;
		const dailyChange = -weeklyLossKg / 7;
		const weights = makeWeightSeries(14, 80, dailyChange);
		const calories = makeCalorieSeries(14, 2000);

		const result = computeAdaptiveTDEE(weights, calories, 14);

		expect(result.estimatedTDEE).not.toBeNull();
		expect(result.estimatedTDEE!).toBeGreaterThan(2400);
		expect(result.estimatedTDEE!).toBeLessThan(2700);
		expect(result.trend).toBe('loss');
		expect(result.avgIntake).toBeCloseTo(2000);
		expect(result.weeklyRate).toBeCloseTo(-0.5, 1);
	});

	test('returns null TDEE with insufficient weight data', () => {
		const weights = makeWeightSeries(3, 80, -0.07);
		const calories = makeCalorieSeries(14, 2000);

		const result = computeAdaptiveTDEE(weights, calories, 14);

		expect(result.estimatedTDEE).toBeNull();
		expect(result.confidence).toBe('insufficient');
	});

	test('returns null TDEE with insufficient calorie data', () => {
		const weights = makeWeightSeries(14, 80, -0.07);
		const calories = makeCalorieSeries(5, 2000);

		const result = computeAdaptiveTDEE(weights, calories, 14);

		expect(result.estimatedTDEE).toBeNull();
		expect(result.confidence).toBe('insufficient');
	});

	test('clamps TDEE below 1200 and sets confidence to low', () => {
		// Extreme gain scenario that would push TDEE very low
		// huge gain + low intake forces clamping
		const weights = makeWeightSeries(14, 60, 1.0); // +7kg/week
		const calories = makeCalorieSeries(14, 1500);

		const result = computeAdaptiveTDEE(weights, calories, 14);

		expect(result.estimatedTDEE).toBe(1200);
		expect(result.confidence).toBe('low');
	});

	test('clamps TDEE above 5000 and sets confidence to low', () => {
		// extreme loss + high intake forces TDEE above 5000
		const weights = makeWeightSeries(14, 100, -1.0); // -7kg/week
		const calories = makeCalorieSeries(14, 4000);

		const result = computeAdaptiveTDEE(weights, calories, 14);

		expect(result.estimatedTDEE).toBe(5000);
		expect(result.confidence).toBe('low');
	});

	test('trend is maintenance when weekly rate is near zero', () => {
		const weights = makeWeightSeries(14, 80, 0);
		const calories = makeCalorieSeries(14, 2200);

		const result = computeAdaptiveTDEE(weights, calories, 14);

		expect(result.trend).toBe('maintenance');
	});

	test('trend is gain when weight is increasing', () => {
		const weights = makeWeightSeries(14, 75, 0.1);
		const calories = makeCalorieSeries(14, 2800);

		const result = computeAdaptiveTDEE(weights, calories, 14);

		expect(result.trend).toBe('gain');
	});
});

describe('detectPlateau', () => {
	test('detects plateau with flat weight', () => {
		const weights = makeWeightSeries(14, 80, 0);
		const calories = makeCalorieSeries(14, 1800);

		const result = detectPlateau(weights, calories, 2200);

		expect(result.isPlateaued).toBe(true);
		expect(result.plateauDays).toBe(14);
		expect(result.estimatedDeficit).toBeCloseTo(400);
		// A deficit with a flat scale is not evidence of adaptation; no cause is asserted.
		expect(result.cause).toBe('none');
	});

	test('does not call a short window a plateau', () => {
		const weights = makeWeightSeries(6, 80, 0);
		const calories = makeCalorieSeries(14, 1800);

		const result = detectPlateau(weights, calories, 2200);

		expect(result.isPlateaued).toBe(false);
	});

	test('sparse weigh-ins are regressed on the date, not the row index', () => {
		// Four points over 13 days losing 0.05 kg/day: a row-index fit would see
		// the same drop compressed into three steps and report ~4x the rate.
		const dense = makeWeightSeries(14, 80, -0.05);
		const sparse = [dense[0], dense[1], dense[2], dense[13]];
		const calories = makeCalorieSeries(14, 2000);

		const tdee = computeAdaptiveTDEE([...sparse, dense[5]], calories, 14);
		expect(tdee.weeklyRate).toBeCloseTo(-0.35, 5);
	});

	test('does not flag plateau during normal weight loss', () => {
		const weights = makeWeightSeries(14, 80, -0.05); // ~0.35kg/week
		const calories = makeCalorieSeries(14, 1800);

		const result = detectPlateau(weights, calories, 2200);

		expect(result.isPlateaued).toBe(false);
	});

	test('returns insufficient confidence with too little data', () => {
		const weights = makeWeightSeries(2, 80, 0);
		const calories = makeCalorieSeries(2, 1800);

		const result = detectPlateau(weights, calories, 2200);

		expect(result.isPlateaued).toBe(false);
		expect(result.confidence).toBe('insufficient');
	});

	test('classifies cause as intake_variance when calorie stddev is high', () => {
		const weights = makeWeightSeries(14, 80, 0);
		const today = new Date();
		today.setUTCHours(0, 0, 0, 0);
		const calories = Array.from({ length: 14 }, (_, i) => {
			const d = new Date(today);
			d.setUTCDate(d.getUTCDate() - (13 - i));
			return { date: d.toISOString().slice(0, 10), calories: i % 2 === 0 ? 1200 : 2800 };
		});

		const result = detectPlateau(weights, calories, 2200);

		expect(result.isPlateaued).toBe(true);
		expect(result.cause).toBe('intake_variance');
	});
});

describe('projectWeight', () => {
	test('projects from the smoothed anchor along a decelerating curve', () => {
		const weights = makeWeightSeries(22, 80, -0.07);
		const weeklyRate = -0.5;

		const result = projectWeight(weights, weeklyRate);

		// Anchor is the trailing 7-day mean, not the last (noisy) raw point.
		const lastSeven = weights.slice(-7).map((w) => w.weightKg);
		expect(result.currentWeight).toBeCloseTo(lastSeven.reduce((s, v) => s + v, 0) / 7, 9);

		// Linear extrapolation is the upper bound on the change; the curve bends
		// as expenditure falls with mass (τ = 7700 / 22 = 350 days).
		const tau = 7700 / 22;
		const expected = (days: number) =>
			result.currentWeight! + (weeklyRate / 7) * tau * (1 - Math.exp(-days / tau));
		expect(result.day30).toBeCloseTo(expected(30), 9);
		expect(result.day60).toBeCloseTo(expected(60), 9);
		expect(result.day90).toBeCloseTo(expected(90), 9);
		const linear90 = result.currentWeight! + (weeklyRate * 90) / 7;
		expect(result.day90!).toBeGreaterThan(linear90);
		expect(result.day90!).toBeLessThan(result.currentWeight!);
		expect(result.confidence).toBe('high');
	});

	test("carries the rate estimate's confidence when given", () => {
		const weights = makeWeightSeries(22, 80, -0.07);
		const result = projectWeight(weights, -0.5, 'low');
		expect(result.confidence).toBe('low');
	});

	test('returns null projections when no weight data', () => {
		const result = projectWeight([], -0.5);

		expect(result.currentWeight).toBeNull();
		expect(result.day30).toBeNull();
		expect(result.day60).toBeNull();
		expect(result.day90).toBeNull();
		expect(result.confidence).toBe('insufficient');
	});

	test('returns low confidence with sparse weight data', () => {
		const weights = makeWeightSeries(10, 80, -0.07);

		const result = projectWeight(weights, -0.5);

		expect(result.confidence).toBe('low');
	});
});

File: mobile/shared/src/commonTest/kotlin/com/bissbilanz/analytics/TdeeTest.kt
package com.bissbilanz.analytics

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

private fun pad2(n: Int): String = n.toString().padStart(2, '0')

class TdeeTest {
    private fun makeWeights(
        dates: List<String>,
        values: List<Double?>,
    ): List<Pair<String, Double?>> = dates.zip(values)

    private fun makeCalories(
        dates: List<String>,
        values: List<Double?>,
    ): List<Pair<String, Double?>> = dates.zip(values)

    @Test
    fun sufficientDataReturnsNonNullTDEE() {
        val dates = (1..20).map { "2024-01-${pad2(it)}" }
        val weights = makeWeights(dates, (1..20).map { 80.0 - it * 0.05 })
        val calories = makeCalories(dates, (1..20).map { 2000.0 })
        val result = computeAdaptiveTDEE(weights, calories)
        assertNotNull(result.estimatedTDEE)
        assertTrue(result.estimatedTDEE!! > 1200)
        assertTrue(result.estimatedTDEE < 5000)
    }

    @Test
    fun insufficientWeightsReturnsNullTDEE() {
        val dates = (1..14).map { "2024-01-${pad2(it)}" }
        val weights = makeWeights(dates.take(4), (1..4).map { 80.0 })
        val calories = makeCalories(dates, (1..14).map { 2000.0 })
        val result = computeAdaptiveTDEE(weights, calories)
        assertNull(result.estimatedTDEE)
        assertEquals(ConfidenceLevel.INSUFFICIENT, result.confidence)
    }

    @Test
    fun insufficientCaloriesReturnsNullTDEE() {
        val dates = (1..14).map { "2024-01-${pad2(it)}" }
        val weights = makeWeights(dates, (1..14).map { 80.0 })
        val calories = makeCalories(dates.take(9), (1..9).map { 2000.0 })
        val result = computeAdaptiveTDEE(weights, calories)
        assertNull(result.estimatedTDEE)
        assertEquals(ConfidenceLevel.INSUFFICIENT, result.confidence)
    }

    @Test
    fun trendClassifiedAsLoss() {
        val dates = (1..20).map { "2024-01-${pad2(it)}" }
        val weights = makeWeights(dates, (1..20).map { 80.0 - it * 0.2 })
        val calories = makeCalories(dates, (1..20).map { 1800.0 })
        val result = computeAdaptiveTDEE(weights, calories)
        assertEquals("loss", result.trend)
        assertTrue(result.weeklyRate < -0.05)
    }

    @Test
    fun trendClassifiedAsGain() {
        val dates = (1..20).map { "2024-01-${pad2(it)}" }
        val weights = makeWeights(dates, (1..20).map { 70.0 + it * 0.2 })
        val calories = makeCalories(dates, (1..20).map { 3000.0 })
        val result = computeAdaptiveTDEE(weights, calories)
        assertEquals("gain", result.trend)
        assertTrue(result.weeklyRate > 0.05)
    }

    @Test
    fun trendClassifiedAsMaintenance() {
        val dates = (1..20).map { "2024-01-${pad2(it)}" }
        val weights = makeWeights(dates, (1..20).map { 75.0 })
        val calories = makeCalories(dates, (1..20).map { 2200.0 })
        val result = computeAdaptiveTDEE(weights, calories)
        assertEquals("maintenance", result.trend)
    }

    @Test
    fun plateauDetectionWithFlatSeries() {
        val dates = (1..14).map { "2024-01-${pad2(it)}" }
        val weights = makeWeights(dates, (1..14).map { 75.0 })
        val calories = makeCalories(dates, (1..14).map { 1900.0 })
        val result = detectPlateau(weights, calories, estimatedTDEE = 2200.0)
        assertTrue(result.isPlateaued)
        assertEquals(14, result.plateauDays)
        // A deficit with a flat scale is not evidence of adaptation; no cause is asserted.
        assertEquals("none", result.cause)
    }

    @Test
    fun plateauNotCalledOnAShortWindow() {
        val dates = (1..6).map { "2024-01-${pad2(it)}" }
        val weights = makeWeights(dates, (1..6).map { 75.0 })
        val calories = makeCalories(dates, (1..6).map { 1900.0 })
        val result = detectPlateau(weights, calories, estimatedTDEE = 2200.0)
        assertTrue(!result.isPlateaued)
    }

    @Test
    fun sparseWeighInsAreRegressedOnTheDate() {
        // Weigh-ins on days 1, 2, 3, 6 and 14 losing 0.05 kg/day: a row-index fit
        // would compress 13 days into four steps and inflate the rate.
        val dates = listOf(1, 2, 3, 6, 14).map { "2024-01-${pad2(it)}" }
        val weights = dates.mapIndexed { _, d -> Pair(d, (80.0 - (d.takeLast(2).toInt() - 1) * 0.05) as Double?) }
        val calories = (1..14).map { Pair("2024-01-${pad2(it)}", 2000.0 as Double?) }
        val result = computeAdaptiveTDEE(weights, calories)
        assertEquals(-0.35, result.weeklyRate, 1e-9)
    }

    @Test
    fun plateauNotDetectedWithSteepSlope() {
        val dates = (1..14).map { "2024-01-${pad2(it)}" }
        val weights = makeWeights(dates, (1..14).map { 80.0 - it * 0.3 })
        val calories = makeCalories(dates, (1..14).map { 1600.0 })
        val result = detectPlateau(weights, calories, estimatedTDEE = 2200.0)
        assertTrue(!result.isPlateaued)
    }

    @Test
    fun insufficientDataForPlateauReturnsNotPlateaued() {
        val dates = listOf("2024-01-01", "2024-01-02")
        val weights = makeWeights(dates, listOf(75.0, 75.0))
        val calories = makeCalories(dates, listOf(2000.0, 2000.0))
        val result = detectPlateau(weights, calories, estimatedTDEE = 2200.0)
        assertTrue(!result.isPlateaued)
        assertEquals(ConfidenceLevel.INSUFFICIENT, result.confidence)
    }

    @Test
    fun weightForecastProjections() {
        val dates = (1..25).map { "2024-01-${pad2(it)}" }
        val weights = makeWeights(dates, (1..25).map { 80.0 - it * 0.1 })
        val result = projectWeight(weights, weeklyRate = -0.5)
        assertNotNull(result.currentWeight)
        assertNotNull(result.day30)
        assertNotNull(result.day60)
        assertNotNull(result.day90)
        assertTrue(result.day30!! < result.currentWeight!!)
        assertTrue(result.day60!! < result.day30)
        assertTrue(result.day90!! < result.day60)
        // The curve flattens as expenditure falls with mass: 90-day loss is under the linear extrapolation.
        assertTrue(result.day90 > result.currentWeight + (-0.5 * 90) / 7)
        assertEquals(ConfidenceLevel.HIGH, result.confidence)
    }

    @Test
    fun weightForecastCarriesTheRateConfidence() {
        val dates = (1..25).map { "2024-01-${pad2(it)}" }
        val weights = makeWeights(dates, (1..25).map { 80.0 - it * 0.1 })
        assertEquals(ConfidenceLevel.LOW, projectWeight(weights, -0.5, ConfidenceLevel.LOW).confidence)
    }

    @Test
    fun weightForecastEmptySeriesReturnsNullCurrentWeight() {
        val result = projectWeight(emptyList(), weeklyRate = -0.5)
        assertNull(result.currentWeight)
        assertNull(result.day30)
        assertNull(result.day60)
        assertNull(result.day90)
        assertEquals(ConfidenceLevel.INSUFFICIENT, result.confidence)
    }

    @Test
    fun weightSeriesWithNullsMixedIn() {
        val weights = (1..20).map { Pair("2024-01-${pad2(it)}", if (it % 3 == 0) null else 80.0 - it * 0.1) }
        val calories = (1..20).map { Pair("2024-01-${pad2(it)}", 2000.0 as Double?) }
        val result = computeAdaptiveTDEE(weights, calories)
        assertTrue(result.sampleSize > 0)
    }

    @Test
    fun singleWeightEntryReturnsInsufficient() {
        val weights = listOf(Pair("2024-01-15", 80.0 as Double?))
        val calories = (1..15).map { Pair("2024-01-${pad2(it)}", 2000.0 as Double?) }
        val result = computeAdaptiveTDEE(weights, calories)
        assertNull(result.estimatedTDEE)
        assertEquals(ConfidenceLevel.INSUFFICIENT, result.confidence)
    }

    @Test
    fun emptyWeightAndCaloriesReturnsInsufficient() {
        val result = computeAdaptiveTDEE(emptyList(), emptyList())
        assertNull(result.estimatedTDEE)
        assertEquals(ConfidenceLevel.INSUFFICIENT, result.confidence)
        assertEquals(0.0, result.avgIntake)
    }

    @Test
    fun projectWeightSingleEntry() {
        val weights = listOf(Pair("2024-01-01", 80.0 as Double?))
        val result = projectWeight(weights, -0.5)
        assertEquals(80.0, result.currentWeight)
        assertEquals(ConfidenceLevel.INSUFFICIENT, result.confidence)
        assertNotNull(result.day30)
    }

    @Test
    fun plateauDetectionSingleWeightReturnsNotPlateaued() {
        val weights = listOf(Pair("2024-01-15", 80.0 as Double?))
        val calories = listOf(Pair("2024-01-15", 2000.0 as Double?))
        val result = detectPlateau(weights, calories, 2000.0)
        assertEquals(false, result.isPlateaued)
        assertEquals(ConfidenceLevel.INSUFFICIENT, result.confidence)
    }
}
